### PR TITLE
fix: remove duplicate enum types

### PR DESCRIPTION
**Pull request by voliva**
We have some enums that they are completely equivalent, yet they have different checksums and different names.

There were two issues with the checksumBuilder, all fixed in this PR to decrease the amount of breaking changes.

## Enums with tuples vs enums with arrays

Enums with variants where one was defined as an `array` (a type of elements with a specific fixed length) was detected as a different type than enums with variants defined with `tuple` with elements of the same type.

For example, `XcmV4Junctions.X2(FixedSizeArray<2, XcmV4Junction>)` and `XcmV3Junctions.X2([XcmV3Junction, XcmV4Junction])` had distinct checksums, but on runtime it's actually the same value: An array of 2 elements of `XcmV3Junctions`.

This has been solved on the lookup resolve function: When it finds an enum variant with type `tuple`, if all of the elements are the same type, it will transform it to an enum variant with type `array`.

## Mirrored types

We also found another case where some circular types that were identical with each other would give different result. As an example, we had in our known-types repo `XcmV2Instruction` and `XcmV2Instruction1`, which were types that when you use them they are absolutely identical, but they had different checksums. And these types were in the polkadot relay chain.

The problem is a bit hard to explain, but it's when there were equivalent types with circular references where at some point you can switch from one to the other. Represented in a graph of types, we have something like this in the metadata:

image

Where `A` and `A'` have the exact same shape, as does `B` and `B'`, `C` and `C'` and `E` and `E'`

This has been solved in the function that calculates the checksum.

Both @josepot and @carlosala helped massively solving this problem. 

# Breaking changes

Given that some of the checksums change, and some types get merged with others, I've taken the chance to review other types and names, and cleanup unused types from the repo. Here's the list of changes:

## Removed

- Remove `DispatchPays`: It's an `Enum<{ Yes: undefined, No: undefined }>` which is also used in other contexts than "DispatchPays". Migrate to `Enum('Yes')` or `Enum('No')`

## Added

- `499e4gfi00j30` => `AssetRatePalletError`: From polkadot and kusama chains. Might collide with preexisting from roccoco and westend
- `8g2s9dnb88h5` => `ProxyEvent`: From assetHub parachains. Might collide with preexisting from relay chains
- `e4o50pgs5mq4e` => `AssetsEvent`: From assetHub
- `e28eun9h9q5hd` => `AssetsPalletCall`: From assetHub
- `fn5p2ma236tv1` => `AssetsPalletError`: From assetHub
- `5o0s7c8q1cc9b` => `PalletError`: From roccoco and westend. Might collide with preexisting from polkadot and kusama
- `69acthsqnju1b` => `XcmPalletCall`: From roccoco and westend. Might collide with preexisting from polkadot and kusama
- `f4pgh470sjqmu` => `XcmPalletError`: From roccoco and westend. Might collide with preexisting from polkadot and kusama
- `84u4ul208g742` => `ReferendaPalletError`: From roccoco and westend. Might collide with preexisting from polkadot and kusama
- `77dda7hps0u37` => `SessionPalletCall`: From assetHub and bridgeHub parachains. Might collide with preexisting from relay chains.

> "Might collide" means that if you have both multiple chains generated, some types with slight variations between chains will have the same name. In that case, the type name will get prefixed with the chain key.

## Merged

- `XcmV4Junctions` into `XcmV3Junctions`
- `XcmVersionedXcm1` into `XcmVersionedXcm`
- `XcmV2Instruction1` into `XcmV2Instruction`
- `XcmV3Instruction1` into `XcmV3Instruction`
- `XcmV4Instruction1` into `XcmV4Instruction`

> The names on the left have been changed to the type names to the right. This just shows the deduplication. 

## Renamed

- `XcmV4Junction` => `XcmV3Junction`
- `XcmV4JunctionNetworkId` => `XcmV3JunctionNetworkId`
- `XcmV3JunctionBodyPart` => `XcmV2JunctionBodyPart`
- `PolkadotRuntimeCommonImplsVersionedLocatableAsset` => `VersionedLocatableAsset`
- `VestingReleases` => `Version`
- `BalancesTypesAdjustmentDirection` => `BalancesAdjustmentDirection`
- `PolkadotPrimitivesV5UpgradeGoAhead` => `UpgradeGoAhead`
- `PolkadotPrimitivesV5UpgradeRestriction` => `UpgradeRestriction`
- `PreimageRequestStatus` => `PreimageOldRequestStatus`
- `StakingPalletConfigOp` => `StakingPalletConfigOpBig`: Collided with the other `ConfigOp`, but this one uses a bignumber as a parameter. This has better naming than "`ConfigOp1`"
- `StakingPalletConfigOp1` => `StakingPalletConfigOp`
- `IdentityTypesData` => `IdentityData`
- `PolkadotPrimitivesV5PvfExecTimeoutKind` => `PvfExecKind`
- `PolkadotPrimitivesV5ValidityAttestation` => `ValidityAttestation`
- `PolkadotPrimitivesV5InvalidDisputeStatementKind` => `InvalidDisputeStatementKind`
- `PolkadotRuntimeParachainsHrmpPalletCall` => `ParachainsHrmpPalletCall`
- `PolkadotPrimitivesV5SlashingSlashingOffenceKind` => `SlashingOffenceKind`
- `PolkadotRuntimeParachainsHrmpEvent` => `ParachainsHrmpEvent`
- `ParachainsDisputesDisputeLocation` => `ParachainsDisputeLocation`
- `ParachainsDisputesDisputeResult` => `ParachainsDisputeResult`
- `PolkadotRuntimeParachainsHrmpPalletError` => `ParachainsHrmpPalletError`
- `PolkadotPrimitivesV5CoreState` => `CoreState`
- `PolkadotPrimitivesV5OccupiedCoreAssumption` => `OccupiedCoreAssumption`
- `PolkadotPrimitivesV5CandidateEvent` => `CandidateEvent`

In general:
- Xcm types have as version number target their lowest compatibility. So you can have XCM v4 use a V2 type (meaning that type is still the same / hasn't changed), but you'll never see a XCM v2 reference any v3 or v4 type.
- Reduce the length of some names that were too specific.

## Cleanup

Types removed, because they are no longer defined in common metadatas.

Note: Don't panic if you see a name here that you were using, as it might still be generated from another checksum. These types are nowhere to be found.

- `TreasuryEvent` (`3ul3cf3die25m`)
- `PreimagePalletCall` (`2vgk5cmoj0b8m`)
- `BalancesPalletCall` (`4hqi6cvv5mjpd`)
- `SessionPalletCall` (`ekmvcec7emmbn`)
- `TreasuryPalletCall` (`conpkk6eh7ho7`)
- `ReferendaPalletCall` (`76ia9hi2mtki0`)
- `NominationPoolsPalletCall` (`77u6cjli3die8`)
- `ParachainsHrmpPalletCall` (`bl5jgc3bginhv`)
- `XcmVersionedXcm1` (`3c4a8k2oha1ud`)
- `XcmV3Instruction1` (`769uhag9pjttp`)
- `NominationPoolsEvent` (`a3rudoq0hsad6`)
- `ParachainsHrmpEvent` (`cd6v2du35hbin`)
- `PreimagePalletError` (`ehqcs6ic5kkkf`)
- `TreasuryPalletError` (`f3i79pgq40jm0`)
- `ReferendaTypesReferendumInfo` (`a7j6l4e7oe7ln`)
- `NominationPoolsPalletError` (`1s3i1fsqjpvr8`)
- `ParachainsParasPvfCheckCause` (`8dtf4dr89ksvh`)
- `ParachainsHrmpPalletError` (`fvc08hks5b16p`)
- `WestendRuntimeRuntimeEvent` (`2dtccotemif6i`)
- `StakingEvent` (`1vpsle7nk6bbf`)
- `WestendRuntimeRuntimeCall` (`3buks4t5p9jo5`)
- `StakingPalletCall` (`4q47o0id5531b`)
- `UtilityPalletCall` (`7ai6hngrgci9q`)
- `RecoveryPalletCall` (`9gk8gt7ihk77k`)
- `SchedulerPalletCall` (`afl9u0rriejnl`)
- `SudoPalletCall` (`c8ick5jm5228d`)
- `ProxyPalletCall` (`f3ubj8sbog6fs`)
- `MultisigPalletCall` (`f5k05q35kbr90`)
- `WhitelistPalletCall` (`4iujedjoq34sc`)
- `PolkadotRuntimeParachainsConfigurationPalletCall` (`5bn08pbgv7o6l`)
- `XcmPalletCall` (`depofgn4a4b4b`)
- `PalletError` (`3bjnl897tu8j2`)
- `StakingPalletError` (`bklnaqjbd0mho`)
- `PolkadotRuntimeParachainsAssignerOnDemandPalletError` (`5kmpac1lo149t`)
- `WestendRuntimeRuntimeError` (`8d4epped9g7uv`)

### CODE REVIEW

File: packages/metadata-builders/tests/checksum-builder.spec.ts [hunk 1]
```diff
@@ -58,7 +65,7 @@ describe("getChecksumBuilder properties", () => {
     )
   })
 
-  it("gives the same result regardless of entry point", () => {
+  it.skip("gives the same result regardless of entry point", () => {
```
Comment (carlosala) [inline packages/metadata-builders/tests/checksum-builder.spec.ts:68]: We should maintain this test, it kinda assures that all checksums are created properly. 👍🏻

File: packages/metadata-builders/tests/checksum-builder.spec.ts [hunk 2]
```diff
@@ -239,6 +247,90 @@ describe("getChecksumBuilder properties", () => {
...
+    })
+    const cDef = lookup[cId].def
+    if (cDef.tag === "composite") {
+      cDef.value[0].type = eId
+    }
+    const builder = getChecksumBuilder({ lookup } as any)
+
+    expect(builder.buildDefinition(aId)).toEqual(builder.buildDefinition(cId))
+    expect(builder.buildDefinition(dId)).toEqual(builder.buildDefinition(eId))
+  })
```
Comment (carlosala) [inline packages/metadata-builders/tests/checksum-builder.spec.ts:332]: Amazing 🎉

**Review by carlosala — APPROVED** [review]

**Comment by carlosala** [comment]
My only concern is if we should maintain `V3` or `V4` for Xcm known types. It looks like migrating everything to the latest one sounds reasonable, and only keeping V2 and V3 in case they differ from V4.
Taking this approach, though, adds a future breaking change in terms of naming in case V5, V6, etc, get released.
The other approach is to show the oldest one, and only show something when they differ. Both seem reasonable to be, but it might be more future-proof to take the latter.

**Comment by carlosala** [comment]
I'd also consider this moment to think if we want to reduce the number of known types that we have.
I feel like many of them are just too much sugar for the goal of the library.

**Comment by josepot** [comment]
> I'd also consider this moment to think if we want to reduce the number of known types that we have. I feel like many of them are just too much sugar for the goal of the library.

This PR already does a significant cleanup. I'm not opposed to the idea of removing some of the unnecessary sugar, but I think that work belongs into a separate PR.

**Comment by carlosala** [comment]
> This PR already does a significant cleanup. I'm not opposed to the idea of removing some of the unnecessary sugar, but I think that work belongs into a separate PR.

I mainly propose to bundle that in just one breaking release 👍🏻

File: packages/metadata-builders/tests/checksum-builder.spec.ts (hunk 1 above)
Comment (carlosala) [reply packages/metadata-builders/tests/checksum-builder.spec.ts:68]: Sounds fair 👍🏻

**Review by josepot — APPROVED** [review]
this is a freaking masterpiece! thank you so much!

**Pull request merged**